src/main/java/jcifs/netbios/UniAddress.java

```diff
@@ -0,0 +1,292 @@
+     * return name;
+     * } catch (NameNotFoundException nnfe) {
+     * uhe = new UnknownHostException(nnfe.getMessage());
+     * }
+     * int dot = name.indexOf('.');
+     * if (dot == -1)
+     * break;
+     * name = name.substring(dot + 1);
+     * }
+     * } catch (NamingException ne) {
+     * if (log.level > 1)
+     * ne.printStackTrace(log);
+     * }
+     * 
+     * throw uhe != null ? uhe : new UnknownHostException("invalid name");
+     * }
+     */
+
+    Object addr;
+    String calledName;
+
+
+    /**
+     * Create a <tt>UniAddress</tt> by wrapping an <tt>InetAddress</tt> or
+     * <tt>NbtAddress</tt>.
+     * 
+     * @param addr
+     *            wrapped address
+     */
+    public UniAddress ( Object addr ) {
+        if ( addr == null ) {
+            throw new IllegalArgumentException();
+        }
+        this.addr = addr;
+    }
+
+
+    /**
+     * Return the IP address of this address as a 32 bit integer.
+     */
+
+    @Override
+    public int hashCode () {
+        return this.addr.hashCode();
+    }
+
+
+    /**
+     * Compare two addresses for equality. Two <tt>UniAddress</tt>s are equal
+     * if they are both <tt>UniAddress</tt>' and refer to the same IP address.
+     */
+    @Override
+    public boolean equals ( Object obj ) {
+        return obj instanceof UniAddress && this.addr.equals( ( (UniAddress) obj ).addr);
+    }
+
+
+    /**
+     * Guess first called name to try for session establishment. This
+     * method is used exclusively by the <tt>jcifs.smb</tt> package.
+     * 
+     * @return the guessed name
+     */
+    @Override
+    public String firstCalledName () {
+        if ( this.addr instanceof NbtAddress ) {
+            return ( (NbtAddress) this.addr ).firstCalledName();
+        }
+
+        this.calledName = ( (InetAddress) this.addr ).getHostName();
+        if ( isDotQuadIP(this.calledName) ) {
+            this.calledName = NbtAddress.SMBSERVER_NAME;
+        }
+        else {
+            int i = this.calledName.indexOf('.');
+            if ( i > 1 && i < 15 ) {
+                this.calledName = this.calledName.substring(0, i).toUpperCase();
+            }
+            else if ( this.calledName.length() > 15 ) {
+                this.calledName = NbtAddress.SMBSERVER_NAME;
+            }
+            else {
+                this.calledName = this.calledName.toUpperCase();
+            }
+        }
+
+        return this.calledName;
+    }
+
+
+    /**
+     * Guess next called name to try for session establishment. This
+     * method is used exclusively by the <tt>jcifs.smb</tt> package.
+     * 
+     * @param tc
+     *            context to use
+     * 
+     * @return guessed alternate name
+     */
+    @Override
+    public String nextCalledName ( CIFSContext tc ) {
+        if ( this.addr instanceof NbtAddress ) {
+            return ( (NbtAddress) this.addr ).nextCalledName(tc);
+        }
+        else if ( this.calledName != NbtAddress.SMBSERVER_NAME ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.